### PR TITLE
Add QR code scanner.

### PR DESCRIPTION
**Pull request by suhailsaqan**
Made it way easier to add friends on nostr.

https://github.com/damus-io/damus/assets/43693074/3f139982-2f37-41b2-bb81-7f817bdd9773

### CODE REVIEW

**Comment by jb55** [comment]
very cool

File: damus/Views/QRCodeView.swift [hunk 1]
```diff
@@ -22,79 +34,215 @@ struct QRCodeView: View {
...
+                .frame(width: 300, height: 300)
+                .cornerRadius(10)
+                .overlay(RoundedRectangle(cornerRadius: 10).stroke(Color("DamusWhite"), lineWidth: 5.0))
+                .overlay(RoundedRectangle(cornerRadius: 10).trim(from: 0.0, to: outerTrimEnd).stroke(Color("DamusPurple"), lineWidth: 5.5)
+                    .rotationEffect(.degrees(-90)))
+                .shadow(radius: 10)
+            
+            Spacer()
+            
+            profileView.animation(.easeInOut(duration: 0.3))
```
Comment (jb55) [inline damus/Views/QRCodeView.swift:235]: is says this is deprecated

File: damus/Views/QRCodeView.swift [hunk 2]
```diff
@@ -7,12 +7,24 @@
...
 
 struct QRCodeView: View {
     let damus_state: DamusState
     
     @Environment(\.dismiss) var dismiss
     @Environment(\.presentationMode) var presentationMode
+    
+    @State var scanResult: Search? = nil
+    
+    @State var profileView: NavigationLink<VStack<TupleView<(_ConditionalContent<ProfilePicView, Image>, Text?, Text?)>>, ProfileView>? = nil
```
Comment (jb55) [inline damus/Views/QRCodeView.swift:20]: this doesn't look right... why is a navigation link in a state variable?

**Comment by jb55** [comment]
This is a bit too messy for me to merge atm :(

**Comment by suhailsaqan** [comment]
@jb55 cleaned it up lol

**Comment by jb55** [comment]
taking another look

**Comment by jb55** [comment]
code looks a lot better!

**Comment by jb55** [comment]
works great, but can we keep the previous design?

File: damus/Views/QRCodeView.swift [hunk 3]
```diff
@@ -7,13 +7,27 @@
...
     let damus_state: DamusState
     @State var pubkey: String
     
     @Environment(\.dismiss) var dismiss
     @Environment(\.presentationMode) var presentationMode
+    
+    @State var scanResult: Search? = nil
+    
+    @State var showProfileView: Bool = false
+    @State var dst: ProfileView? = nil
```
Comment (jb55) [inline damus/Views/QRCodeView.swift:22]: views should not be state variables

File: damus/Views/QRCodeView.swift [hunk 4]
```diff
@@ -23,80 +37,198 @@ struct QRCodeView: View {
...
+                        switch scanResult {
+                            case .profile(let prof):
+                                if (currentScannedCode != prof) {
+                                    generator.impactOccurred()
+                                    cameraAnimate()
+                                }
+                                currentScannedCode = prof
+                                let decoded = try? bech32_decode(prof)
+                                let hex = hex_encode(decoded!.data)
+                                profile = damus_state.profiles.lookup(id: hex)
```
Comment (jb55) [inline damus/Views/QRCodeView.swift:181]: The proper way to do this is to make a pubkey state variable instead of the actual view

**Review by jb55 — CHANGES_REQUESTED** [review]

**Comment by BenGWeeks** [comment]
Propose that rather than sliding left/right, as it might not be immediately obvious, we have a simple "Scan" button that takes you to the Scan page and a "My Code" on the Scan page that takes you back. This is how WhatsApp works.

FYI @SashaPavlova @robagreda should probably be in the designs too.

**Comment by BenGWeeks** [comment]
closes [#1149](https://github.com/damus-io/damus/issues/1149)

**Comment by jb55** [comment]
On Sun, May 28, 2023 at 08:22:10AM -0700, Ben Weeks wrote:
>Propose that rather than sliding left/right, as it might not be immediately obvious, we have a simple "Scan" button that takes you to the Scan page and a "My Code" on the Scan page that takes you back. This is how WhatsApp works.
>
>FYI @SashaPavlova @robagreda should probably be in the designs too.

Agreed, the sliding thing is weird to me

**Comment by BenGWeeks** [comment]
Hadn't realized @robagreda had designed this screen up actually. You can find it here:

CleanShot 2023-05-30 at 16 12 40

Ref: https://www.figma.com/file/ORaT1T0Ywfbm0sIjwy5Rgq/Damus-iOS?type=design&node-id=0%3A1&t=eCerbHhgYI5tLFQt-1

**Comment by jb55** [comment]
Thanks! I merged this in the current state, still needs redesign. @suhailsaqan

**Pull request closed without merging**